<a href="https://colab.research.google.com/github/marisbotero/Master_IA/blob/main/readPars_Maris_Botero.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Paráfrasis

**Tarea 1**

In [19]:
import pandas as pd

In [20]:
#dt = pd.read_csv('./test-gold/STS.input.SMTeuroparl.txt',sep='\t',header=None)
dt = pd.read_csv('STS.input.SMTeuroparl.txt',sep='\t',header=None)

In [21]:
dt.head()

0  \
0  The leaders have now been given a new chance a...   
1  Amendment No 7 proposes certain changes in the...   
2  Let me remind you that our allies include ferv...   
3        The vote will take place today at 5.30 p.m.   
4  The fishermen are inactive, tired and disappoi...   

                                                   1  
0  The leaders benefit aujourd' hui of a new luck...  
1  Amendment No 7 is proposing certain changes in...  
2  I would like to remind you that among our alli...  
3                 The vote will take place at 5.30pm  
4  The fishermen are inactive, tired and disappoi...

In [22]:
#dt['gs'] = pd.read_csv('./test-gold/STS.gs.SMTeuroparl.txt',sep='\t',header=None)
dt['gs'] = pd.read_csv('STS.gs.SMTeuroparl.txt',sep='\t',header=None)

In [23]:
dt.shape

(459, 3)

In [24]:
dt.head()

0  \
0  The leaders have now been given a new chance a...   
1  Amendment No 7 proposes certain changes in the...   
2  Let me remind you that our allies include ferv...   
3        The vote will take place today at 5.30 p.m.   
4  The fishermen are inactive, tired and disappoi...   

                                                   1    gs  
0  The leaders benefit aujourd' hui of a new luck...  4.50  
1  Amendment No 7 is proposing certain changes in...  5.00  
2  I would like to remind you that among our alli...  4.25  
3                 The vote will take place at 5.30pm  4.50  
4  The fishermen are inactive, tired and disappoi...  5.00

## TODO
1. Compute la similitud entre las oraciones utilizando distancia de Jaccard. Añade una columna "jaccard" a los DataFrames con dicha similitud.
2. Calcula la correlación de Pearson como: <br>
```
from scipy.stats import pearsonr
pearsonr(dt['gs'], dt['jaccard'])[0]
```

In [30]:
import pandas as pd
from sklearn.metrics import jaccard_score
from scipy.stats import pearsonr


# Función para calcular la similitud de Jaccard entre dos frases
def jaccard_similarity(sentence1, sentence2):
    set1 = set(sentence1.lower().split())
    set2 = set(sentence2.lower().split())
    intersection = set1.intersection(set2)
    union = set1.union(set2)
    return len(intersection) / len(union)

# Aplica la función a cada par de oraciones y almacena los resultados en una nueva columna
dt['jaccard'] = dt.apply(lambda row: jaccard_similarity(row[0], row[1]), axis=1)

# Calcula la correlación de Pearson entre la similitud de Jaccard y el Gold Standard
correlation, _ = pearsonr(dt['jaccard'], dt['gs'])
print('Pearsons correlation: %.3f' % correlation)


Pearsons correlation: 0.448


In [31]:
# Define los bordes de los intervalos
bins = [0, 0.25, 0.5, 0.75, 1]

# Define los nombres de las categorías
labels = ['baja', 'moderada', 'alta', 'muy alta']

# Crea la nueva columna
dt['jaccard_categoria'] = pd.cut(dt['jaccard'], bins=bins, labels=labels, include_lowest=True)

In [32]:
dt['jaccard_categoria'].value_counts()

moderada    220
baja        120
muy alta     73
alta         46
Name: jaccard_categoria, dtype: int64

- En el análisis de los datos, se encontró que la mayoría de las oraciones tienen una similitud moderada, con 220 casos clasificados en esta categoría según el índice de Jaccard. Esto sugiere que, en términos generales, las oraciones comparten algunas palabras clave, pero no necesariamente tienen una similitud alta.

- La siguiente categoría más frecuente es la similitud baja, con 120 casos. Estos son pares de oraciones que tienen poca superposición en términos de las palabras que utilizan.

- En contraste, solo 46 casos presentaron una similitud alta y 73 casos presentaron una similitud muy alta. Los casos de alta similitud son aquellos en los que una cantidad significativa de palabras coincidían entre las dos oraciones. Los casos de similitud muy alta, que son aún menos frecuentes, son aquellos donde la superposición de palabras es muy grande, acercándose a frases idénticas.

- Este análisis sugiere que, aunque hay cierta similitud en las oraciones examinadas, esta similitud tiende a ser moderada o baja. Solo un número reducido de casos presenta una alta o muy alta similitud.

- Es importante tener en cuenta que la similitud de Jaccard se basa solamente en las palabras en común y no considera el orden de las palabras ni la estructura de las oraciones, lo cual puede tener un gran impacto en el significado de las oraciones.

Tu coeficiente de correlación de Pearson es de 0.448. Esto indica una correlación moderadamente positiva entre la similitud de Jaccard y las puntuaciones de "Gold Standard". En otras palabras, a medida que la similitud de Jaccard aumenta, la puntuación del "Gold Standard" tiende a aumentar, y viceversa. Sin embargo, la correlación no es fuerte, lo que significa que no todas las veces que la similitud de Jaccard es alta la puntuación del "Gold Standard" es alta también, y viceversa.

**Tarea 2**

In [26]:
import nltk
from sklearn.metrics import jaccard_score
from scipy.stats import pearsonr
from nltk.corpus import stopwords

# Descargar el conjunto de stopwords
nltk.download('stopwords')

# Obtener las stopwords en inglés
stopwords = set(stopwords.words('english'))

# Función para calcular la similitud de Jaccard entre dos frases
def jaccard_similarity(sentence1, sentence2):
    set1 = set(word for word in sentence1.lower().split() if word not in stopwords)
    set2 = set(word for word in sentence2.lower().split() if word not in stopwords)
    intersection = set1.intersection(set2)
    union = set1.union(set2)
    
    return len(intersection) / len(union)

# Supongamos que tienes dos columnas con texto
dt['jaccard'] = dt.apply(lambda row: jaccard_similarity(row[0], row[1]), axis=1)

# Ahora puedes calcular la correlación de Pearson entre la similitud de Jaccard y el "Gold Standard"
correlation, _ = pearsonr(dt['jaccard'], dt['gs'])
print('Pearsons correlation:', correlation)


Pearsons correlation: 0.4531989226806014


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [27]:
dt.head()

0  \
0  The leaders have now been given a new chance a...   
1  Amendment No 7 proposes certain changes in the...   
2  Let me remind you that our allies include ferv...   
3        The vote will take place today at 5.30 p.m.   
4  The fishermen are inactive, tired and disappoi...   

                                                   1    gs   jaccard  
0  The leaders benefit aujourd' hui of a new luck...  4.50  0.333333  
1  Amendment No 7 is proposing certain changes in...  5.00  0.750000  
2  I would like to remind you that among our alli...  4.25  0.166667  
3                 The vote will take place at 5.30pm  4.50  0.428571  
4  The fishermen are inactive, tired and disappoi...  5.00  1.000000

In [28]:
# Define los bordes de los intervalos
bins = [0, 0.25, 0.5, 0.75, 1]

# Define los nombres de las categorías
labels = ['baja', 'moderada', 'alta', 'muy alta']

# Crea la nueva columna
dt['jaccard_categoria'] = pd.cut(dt['jaccard'], bins=bins, labels=labels, include_lowest=True)


In [29]:
dt['jaccard_categoria'].value_counts()

baja        180
moderada    161
muy alta     70
alta         48
Name: jaccard_categoria, dtype: int64

Estos cambios indican que al eliminar las stopwords, que son palabras que se utilizan frecuentemente pero que no aportan mucho al significado de una frase, la similitud entre muchas de las frases disminuyó, llevándolas de la categoría "moderada" a la categoría "baja". En otras palabras, sin las stopwords, las frases parecen ser menos similares entre sí en términos de las palabras que utilizan.

Por otro lado, las frases con similitud alta y muy alta no experimentaron cambios significativos. Esto sugiere que la eliminación de las stopwords no tuvo mucho impacto en las frases que ya eran muy similares entre sí.

**Tarea 3**

In [33]:
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from sklearn.metrics import jaccard_score
from scipy.stats import pearsonr

# Descargar los recursos necesarios
nltk.download('stopwords')
nltk.download('wordnet')

# Obtener las stopwords en inglés
stopwords = set(stopwords.words('english'))

# Inicializar el lematizador
lemmatizer = WordNetLemmatizer()

# Función para calcular la similitud de Jaccard entre dos frases
def jaccard_similarity(sentence1, sentence2):
    set1 = set(lemmatizer.lemmatize(word) for word in sentence1.lower().split() if word not in stopwords)
    set2 = set(lemmatizer.lemmatize(word) for word in sentence2.lower().split() if word not in stopwords)
    intersection = set1.intersection(set2)
    union = set1.union(set2)
    
    return len(intersection) / len(union)

# Calcular la similitud de Jaccard para cada par de frases en el DataFrame
dt['jaccard_lemmas'] = dt.apply(lambda row: jaccard_similarity(row[0], row[1]), axis=1)

# Calcular la correlación de Pearson entre la similitud de Jaccard con lemas y el "Gold Standard"
correlation, _ = pearsonr(dt['jaccard_lemmas'], dt['gs'])
print('Pearsons correlation:', correlation)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...


Pearsons correlation: 0.45946149604335706


In [34]:
dt.head()

0  \
0  The leaders have now been given a new chance a...   
1  Amendment No 7 proposes certain changes in the...   
2  Let me remind you that our allies include ferv...   
3        The vote will take place today at 5.30 p.m.   
4  The fishermen are inactive, tired and disappoi...   

                                                   1    gs   jaccard  \
0  The leaders benefit aujourd' hui of a new luck...  4.50  0.333333   
1  Amendment No 7 is proposing certain changes in...  5.00  0.769231   
2  I would like to remind you that among our alli...  4.25  0.318182   
3                 The vote will take place at 5.30pm  4.50  0.600000   
4  The fishermen are inactive, tired and disappoi...  5.00  1.000000   

  jaccard_categoria  jaccard_lemmas  
0          moderada        0.333333  
1          muy alta        0.750000  
2          moderada        0.166667  
3              alta        0.428571  
4          muy alta        1.000000

In [35]:
# Categorizar la similitud de Jaccard basada en lemas
dt['jaccard_lemmas_categoria'] = pd.cut(dt['jaccard_lemmas'], bins=[0,0.25,0.5,0.75,1], labels=['baja', 'moderada', 'alta', 'muy alta'])

# Ver la distribución de las categorías de similitud
print(dt['jaccard_lemmas_categoria'].value_counts())


moderada    167
baja        135
muy alta     70
alta         48
Name: jaccard_lemmas_categoria, dtype: int64


- Los resultados indican que después de aplicar la lematización en el cálculo de la similitud de Jaccard, la mayor cantidad de pares de frases cayó en la categoría de "moderada" (167 frases), seguida de "baja" (135 frases), "muy alta" (70 frases) y "alta" (48 frases).

- Comparado con los resultados anteriores (antes de aplicar la lematización), se puede observar que la distribución de las categorías ha cambiado. Concretamente, la cantidad de frases con similitud "moderada" ha aumentado, mientras que la cantidad de frases con similitud "baja" ha disminuido. Esto indica que la lematización ha incrementado la similitud percibida entre los pares de frases al considerar las formas base de las palabras, en lugar de sus formas exactas.

- Estos resultados pueden sugerir que la lematización podría ser útil para mejorar la comparación de textos al tratar diferentes formas de una palabra como la misma. Sin embargo, también es importante recordar que la lematización puede no ser siempre apropiada, dependiendo del contexto y de los matices de significado que pueden estar asociados con las diferentes formas de una palabra

¿Qué es mejor: palabras o lemas?
los resultados muestran que la distribución de las categorías de similitud cambió después de la lematización. Esto sugiere que la lematización pudo haber capturado la similitud entre las frases de una manera un poco diferente a cuando se utilizaron las palabras exactas.

¿Crees que podría funcionar mejor para cualquier par de textos?
No necesariamente. La utilidad de la lematización  puede variar ampliamente dependiendo del conjunto de datos y del problema específico que estés tratando de resolver. La lematización tiende a ser más útil en situaciones donde las palabras se utilizan en una variedad de formas y quieres reducir estas formas a su base para simplificar el análisis.Sin embargo, hay situaciones donde la lematización puede no ser la mejor opción. Es bueno probar diferentes enfoques y ver cual se adapta mejor al problema.